libtransact/src/state/btree.rs

```diff
@@ -0,0 +1,181 @@
+            .indexes
+            .get(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .get(key)
+        {
+            Some(value) => Ok(Some(value.to_vec())),
+            None => Ok(None),
+        }
+    }
+
+    /// Returns a cursor against the main database. The cursor iterates over
+    /// the entries in the natural key order.
+    fn cursor(&self) -> Result<LmdbDatabaseReaderCursor, DatabaseError> {
+        unimplemented!()
+    }
+
+    /// Returns a cursor against the given index. The cursor iterates over
+    /// the entries in the index's natural key order.
+    fn index_cursor(&self, index: &str) -> Result<LmdbDatabaseReaderCursor, DatabaseError> {
+        unimplemented!()
+    }
+
+    /// Returns the number of entries in the main database.
+    fn count(&self) -> Result<usize, DatabaseError> {
+        unimplemented!()
+    }
+
+    /// Returns the number of entries in the given index.
+    fn index_count(&self, index: &str) -> Result<usize, DatabaseError> {
+        unimplemented!()
+    }
+}
+//
+pub struct BTreeWriter {
+    db: Arc<Mutex<BTreeDatabase>>,
+}
+
+impl DatabaseWriter for BTreeWriter {
+    fn put(&mut self, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
+        if self.db.lock().expect("error").main.contains_key(key) {
+            return Err(DatabaseError::DuplicateEntry);
+        }
+        self.db
+            .lock()
+            .expect("error")
+            .main
+            .insert(key.to_vec(), value.to_vec());
+        Ok(())
+    }
+
+    fn overwrite(&mut self, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
+        self.db
+            .lock()
+            .expect("error")
+            .main
+            .insert(key.to_vec(), value.to_vec());
+        Ok(())
+    }
+
+    fn index_put(&mut self, index: &str, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
+        self.db
+            .lock()
+            .expect("error")
+            .indexes
+            .get_mut(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .insert(key.to_vec(), value.to_vec());
+        Ok(())
+    }
+
+    fn delete(&mut self, key: &[u8]) -> Result<(), DatabaseError> {
+        self.db
+            .lock()
+            .expect("error")
+            .main
+            .remove(key)
+            .ok_or(DatabaseError::WriterError("Key not found".to_string()))?;
+        Ok(())
+    }
+
+    fn index_delete(&mut self, index: &str, key: &[u8]) -> Result<(), DatabaseError> {
+        self.db
+            .lock()
+            .expect("error")
+            .indexes
+            .get_mut(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .remove(key)
+            .ok_or(DatabaseError::WriterError("Key not found".to_string()))?;
+        Ok(())
+    }
+
+    fn commit(&mut self) -> Result<(), DatabaseError> {
```

Comment:
I was unsure what to do with `commit` here. I understand that until we call `commit` we don't want any of the changes applied by calling the other methods to persist, correct? And if any of the transactions fail, we want all of them to fail? Any suggestions on the best way to deal with this?